examples/graphics/mesh-decals.html

```diff
@@ -0,0 +1,245 @@
+            var g = Math.random() * 255;
+            var b = Math.random() * 255;
+
+            var j;
+            for (j = 0; j < 4; j++) {
+                this.colors[i * 16 + j * 4 + 0] = r;
+                this.colors[i * 16 + j * 4 + 1] = g;
+                this.colors[i * 16 + j * 4 + 2] = b;
+                this.colors[i * 16 + j * 4 + 3] = 0;    // alpha is not used by shader
+            }
+
+            // vertex positions to form a square quad with random rotation and size
+            this.positions[12 * i + 0] = pos.x + size * Math.sin(angle);      this.positions[12 * i + 1] = 0;     this.positions[12 * i + 2] = pos.z + size * Math.cos(angle);      angle += Math.PI * 0.5;
+            this.positions[12 * i + 3] = pos.x + size * Math.sin(angle);      this.positions[12 * i + 4] = 0;     this.positions[12 * i + 5] = pos.z + size * Math.cos(angle);      angle += Math.PI * 0.5;
+            this.positions[12 * i + 6] = pos.x + size * Math.sin(angle);      this.positions[12 * i + 7] = 0;     this.positions[12 * i + 8] = pos.z + size * Math.cos(angle);      angle += Math.PI * 0.5;
+            this.positions[12 * i + 9] = pos.x + size * Math.sin(angle);      this.positions[12 * i + 10] = 0;    this.positions[12 * i + 11] = pos.z + size * Math.cos(angle);     angle += Math.PI * 0.5;
+        }
+
+        // Set an update function on the app's update event
+        var self = this;
+        var time = 0;
+        var firstTime = true;
+        var decalIndex = 0;
+        app.on("update", function (dt) {
+
+            var previousTime = time;
+            time += dt;
+
+            // Bounce the ball around in a circle with changing radius
+            var radius = Math.abs(Math.sin(time * 0.55) * 9);
+            var previousElevation = 2 * Math.cos(previousTime * 7);
+            var elevation = 2 * Math.cos(time * 7);
+            self.ball.setLocalPosition(new pc.Vec3(radius * Math.sin(time), 0.5 + Math.abs(elevation), radius * Math.cos(time)));
+
+            // When ball crossed the ground plane
+            var positionsUpdated = false;
+            var colorsUpdated = false;
+            if ((previousElevation < 0 && elevation >= 0) || (elevation < 0 && previousElevation >= 0)) {
+
+                positionsUpdated = true;
+                colorsUpdated = true;
+                
+                // create new decal at next index, and roll the index around if out of range
+                CreateDecal(decalIndex, self.ball.getLocalPosition());
+                decalIndex++;
+                if (decalIndex >= self.numDecals)
+                    decalIndex = 0;
+            }
+
+            // fade out all vertex colors once a second
+            if (Math.round(time) != Math.round(previousTime)) {
+                for (i = 0; i < self.colors.length; i++)
+                    self.colors[i] -= 2;
+
+                colorsUpdated = true;
+            }
+
+            // on the first frame
+            if (firstTime) {
```

Comment:
MeshInstance cannot be created with mesh that does not have VertexBuffer and VertexFormat yet, as it does this initialization. 

`        this._shaderDefs |= mesh.vertexBuffer.format.hasUv0 ? pc.SHADERDEF_UV0 : 0;
        this._shaderDefs |= mesh.vertexBuffer.format.hasUv1 ? pc.SHADERDEF_UV1 : 0;
        this._shaderDefs |= mesh.vertexBuffer.format.hasColor ? pc.SHADERDEF_VCOLOR : 0;
        this._shaderDefs |= mesh.vertexBuffer.format.hasTangents ? pc.SHADERDEF_TANGENTS : 0;
`
So I could either:
- change this requirement and allow mesh to be assigned later .. it didn't seem like ideal option
- do fake mesh update with vertex streams at init time to build it's format up. But that would not work in general case when you call .Clear on mesh and set it up with different format.
- do what I did .. create MeshInstance and the rest only when Mesh was generated.

Reply:
Why can't you create the vertex and index buffer upfront? I was thinking you can do all initialization (including async download of texture etc), create meshes, instances, materials and then when it's all ready you call app.start(). At this point update() will start being called and your code can assume everything is in place.

I think if it is possible to separate out the init and update, the code will be a little easier to follow, but also the engine-visual-test will work better this way.

Though, of course I may have missed something and it's not a straightforward change.

Reply:
fair point .. I was trying to avoid updating mesh unless needed .. but clarity is not ideal .. I'm changing these to initialize vertex buffer upfront.

Reply:
updated